endpoint.go

```diff
@@ -0,0 +1,58 @@
+// Copyright (C) 2021 ScyllaDB
+
+package gocql
+
+import (
+	"crypto/tls"
+	"fmt"
+	"net"
+)
+
+// Endpoint aggregates all necessary values needed to connect to particular node in the cluster.
+type Endpoint struct {
+	// Server is an address where connection should be dialed.
+	Server string
+	// ProxyURL is an address to SOCKS5 proxy through which connections should go through.
+	ProxyURL string
+	// TLSConfig is a TLS Configuration required to connect to particular node.
+	TLSConfig *tls.Config
+}
+
+// EndpointFactory creates an Endpoint from system.local/peer table rows.
+type EndpointFactory interface {
+	Create(row map[string]interface{}) (*Endpoint, error)
+}
+
+// DirectEndpointFactory returns Endpoints going directly to particular node.
+// Note: TLS configuration is shared between entire cluster and it's taken from ClusterConfig.SslOpts.
+type DirectEndpointFactory struct {
+	clusterConfig *ClusterConfig
+}
+
+func (f *DirectEndpointFactory) Create(row map[string]interface{}) (*Endpoint, error) {
```

Comment:
How does that relate to 

```
func (h *HostInfo) connectAddressLocked() (net.IP, string) {
	if validIpAddr(h.connectAddress) {
		return h.connectAddress, "connect_address"
	} else if validIpAddr(h.rpcAddress) {
		return h.rpcAddress, "rpc_adress"
	} else if validIpAddr(h.preferredIP) {
		// where does perferred_ip get set?
		return h.preferredIP, "preferred_ip"
	} else if validIpAddr(h.broadcastAddress) {
		return h.broadcastAddress, "broadcast_address"
	} else if validIpAddr(h.peer) {
		return h.peer, "peer"
	}
	return net.IPv4zero, "invalid"
}
```

Reply:
IMO it always returns value of rpc_address, but maybe i don't know all possible configuration options. I'll add similar logic here to make sure i won't break some edge case.